dimred/src/gwas/train_val_split.py

```diff
@@ -1,224 +1,285 @@
+            phenotype_path (str): Path with only phenotype data
+
+        """   
+        pca = pandas.read_table(pca_path)
+        cov_pheno = pandas.read_table(cov_pheno_path)
+        cov_columns = list(cov_pheno.columns)[2:-1]
+        pheno_column = cov_pheno.columns[-1]
+        # print(f'COV_PHENO_COLUMNS are: {cov_columns}')
+        merged = pca.merge(cov_pheno, how='inner', on=['FID', 'IID'])
+        
+        pca_cov = merged.loc[:, ['FID', 'IID'] + [f'PC{i}' for i in range(1, 11)] + cov_columns]
+        pca_cov.fillna(pca_cov.mean(), inplace=True)
+        phenotype = merged.loc[:, ['FID', 'IID'] + [pheno_column]]
+
+        phenotype.to_csv(phenotype_path, sep='\t', index=False)
+        pca_cov.to_csv(pca_cov_path, sep='\t', index=False)
+
+
+    def standardize_covariates(self, node_index: int, covariates: List[str] = None):
+        """Z-standardizes {covariates} for each fold and particular node {node_index}
+
+        Args:
+            node_index (int): Index of node
+            covariates (List[str], optional): Covariates to standardize. If None, every covariate will be standardized. Defaults to None.
+        """        
+        for fold_index in range(FOLD_COUNT):
+            self.standardize(
+                    self.split.get_pca_cov_path(node_index, fold_index, 'train'),
+                    self.split.get_pca_cov_path(node_index, fold_index, 'val'),
+                    self.split.get_pca_cov_path(node_index, fold_index, 'test'),
+                    covariates
+            )
+
+            # standardize phenotype in phenotype-only file
+            '''
+            self.standardize(
+                    self.split.get_phenotype_path(node_index, fold_index, 'train'),
+                    self.split.get_phenotype_path(node_index, fold_index, 'val'),
+                    self.split.get_phenotype_path(node_index, fold_index, 'test'),
+                    None
+            )
+            '''
+
+    def standardize(self, train_path: str, val_path: str, test_path: str, columns: List[str]):
+        """
+        Infers mean and std from columns in {train_path} and standardizes both train, test and val columns in-place
+        TODO: think about non-iid data!!!
+
+        Args:
+            train_path (str): Path to .tsv file with train data. First two columns should be FID, IID
+            val_path (str): Path to .tsv file with val data. First two columns should be FID, IID
+            columns (List[str]): List of columns to standardize. By default all columns except FID and IID will be standardized. 
+        """    
+        train_data = pandas.read_table(train_path)
+        val_data = pandas.read_table(val_path)
+        test_data = pandas.read_table(test_path)
+
+        scaler = StandardScaler()
+        if columns is None:
+            train_data.iloc[:, 2:] = scaler.fit_transform(train_data.iloc[:, 2:]) # 0,1 are FID, IID
+            val_data.iloc[:, 2:] = scaler.transform(val_data.iloc[:, 2:])
```

Comment:
It's generally recommended to normalise test and val datasets using mean and variance from the train dataset instead of calculation mean and variance in-house
https://stackoverflow.com/questions/49444262/normalize-data-before-or-after-split-of-training-and-testing-data

Reply:
scaler.transform uses mean and variance obtained from train_data using fit_transform method